### ISSUE TITLE
DistributedDataParallel doesn't converge well when using MPI

### ISSUE THREAD

**Issue by apaszke** [pytorch/pytorch#4406]
## Originally posted by @alsrgv:

I got Gloo working with 4 nodes with 4 GPUs each, but it fails on 32 nodes.  I tried MPI backend with `DistributedDataParallel` and found that it doesn't converge very well.

I scaled back experiment to 4 nodes with 4 GPUs each and observed that MPI backend with `DistributedDataParallel` indeed converges much worse with the same batch size & lr.  On the chart below, gray line is Gloo and green line is MPI. 

<img width="714" alt="screen shot 2017-12-27 at 4 03 34 pm" src="https://user-images.githubusercontent.com/16640218/34396423-934bfc4e-eb1f-11e7-8fca-8bd3a9ed8b33.png">

I was finally able to run large-scale experiment with doing a simple gradient averaging instead of using `DistributedDataParallel` functionality:

```
def average_gradients(model):
    size = float(dist.get_world_size())
    for param in model.parameters():
        dist.all_reduce(param.grad.data, op=dist.reduce_op.SUM)
        param.grad.data /= size
```

I did two more runs - one with `DataParallel` and one worker per machine (blue line), and another is with one worker per GPU (orange line).

<img width="707" alt="screen shot 2017-12-27 at 4 21 27 pm" src="https://user-images.githubusercontent.com/16640218/34396602-059783b6-eb22-11e7-9ba1-079bcbbe5a04.png">

Unfortunately, it's slow.  Any idea how to make `DistributedDataParallel` work correctly with MPI?

**Comment by SsnL** [comment]
cc @teng-li @ailzhang

**Comment by xhzhao** [comment]
@apaszke The document shows that nn.DistributedDataParallel only support nccl and gloo. My question is: **how could i use nn.DistributedDataParallel  with MPI backend?**

BTW, mingfei get some points about the convergency issue. @mingfeima

**Comment by mingfeima** [comment]
Should gradients be accumulated across ranks or averaged?
From my experience, averaging is likely to slow the converging since gradient is smaller than normal.
 
[TorchMPI](https://github.com/facebookresearch/TorchMPI/blob/master/torchmpi/nn.lua#L49) allreduces the gradients:
```lua
-- Synchronize gradients
M.synchronizeGradients = function(net)
   if not net.parameters then return end
   local p, g = net:parameters()
   for i, gw in ipairs(g) do
      local allreduceTensor = selectCollective(gw, 'sync', 'allreduceTensor')
      allreduceTensor(gw)
   end
end
```
[distlearn](https://github.com/twitter/torch-distlearn/blob/master/lua/AllReduceSGD.lua#L18) averages the gradients:
```lua
   -- Sum and normalize the gradients of all nodes
   local function sumAndNormalizeGradients(grads)
      -- All reduce and sum the gradients
      local _,n = tree.allReduce(grads, function(a, b) return a:add(b) end)
      -- Normalize them by the # of nodes that contributed
      -- Not all nodes contribute to every step due to uneven partitioning of data
      if n > 1 then
         tree.walkTable(grads, function(grad)
            grad:mul(1/n)
         end)
      end
      -- This node contributed to this step
      stepsPerNode[tree.nodeIndex] = stepsPerNode[tree.nodeIndex] + 1
   end
```

**Comment by apaszke** [comment]
They should be averages

**Comment by pengjiang030** [comment]
I have met the same problem. It seems distributed training with gradient averaging is exactly the same as minibatch on a single node; however, the algorithm converges slower than single node training. I looked at the code in DistributedDataParallel, it actually synchronizes the parameters at the beginning of forward propagation.

In Source code for torch.nn.parallel.distributed:
```
def forward(self, *inputs, **kwargs):
        self.need_reduction = True
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
        self._sync_params()
        if len(self.device_ids) == 1:
            return self.module(*inputs[0], **kwargs[0])
        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
        return self.gather(outputs, self.output_device)
```

The gradient averaging is registered:
 ```
 # Now register the reduction hook on the parameters
        for p in self.module.parameters():
            if not p.requires_grad:
                continue

            def allreduce_hook(*unused):
                Variable._execution_engine.queue_callback(reduction_fn_nccl)

            p.register_hook(allreduce_hook)

```
and executed later at the end of backward propagation.

I wonder if it is some internal state on different nodes that is not communicated with gradient averaging causes the slow down of convergence. Is the parameter averaging at the beginning of each iteration really necessary?

**Comment by xhzhao** [comment]
@pengjiang030 we proposed a [PR ](https://github.com/pytorch/pytorch/pull/5919)to enable DDP-CPU, would you like to have a try with this new interface on CPU side?
DDP-CPU code: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed_cpu.py 
BTW, the DDP-CPU interface is more simple then the DDP on GPU.

**Comment by dashesy** [comment]
@pengjiang030 did you find an answer, and if `_sync_params` is always necessary? It may be of interest to people interested in the distributed trainings [to discuss](https://discuss.pytorch.org) the findings in performance improvements. One point I think to consider: there is a single process per GPU, or single process per node when you create the `world_size` and DDP works in both cases. This line:

    inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)

only should matter if there are multiple GPUs in each node, DDP only parallelizes among nodes/processes. I think it is worthwhile to always spawn a single process per GPU (e.g. using `CUDA_VISIBLE_DEVICES` or some other method).
